File: Saniya/ads/quadtree.c
#include <stdio.h>
#include <stdlib.h>

#define CAPACITY 4

// Point structure
typedef struct {
    int x, y;
} Point;

// Rectangle boundary
typedef struct {
    int x, y;      // center
    int w, h;      // half width & height
} Boundary;

// Quadtree node
typedef struct Quadtree {
    Boundary boundary;
    Point points[CAPACITY];
    int count;
    int divided;

    struct Quadtree *nw, *ne, *sw, *se;
} Quadtree;

// Create Quadtree node
Quadtree* createQuadtree(Boundary boundary) {
    Quadtree* qt = (Quadtree*)malloc(sizeof(Quadtree));
    qt->boundary = boundary;
    qt->count = 0;
    qt->divided = 0;
    qt->nw = qt->ne = qt->sw = qt->se = NULL;
    return qt;
}

// Check if point inside boundary
int contains(Boundary b, Point p) {
    return (p.x >= (b.x - b.w) &&
            p.x <= (b.x + b.w) &&
            p.y >= (b.y - b.h) &&
            p.y <= (b.y + b.h));
}

// Subdivide
void subdivide(Quadtree* qt) {
    int x = qt->boundary.x;
    int y = qt->boundary.y;
    int w = qt->boundary.w / 2;
    int h = qt->boundary.h / 2;

    Boundary nw = {x - w, y - h, w, h};
    Boundary ne = {x + w, y - h, w, h};
    Boundary sw = {x - w, y + h, w, h};
    Boundary se = {x + w, y + h, w, h};

    qt->nw = createQuadtree(nw);
    qt->ne = createQuadtree(ne);
    qt->sw = createQuadtree(sw);
    qt->se = createQuadtree(se);

    qt->divided = 1;
}

// Insert point
int insert(Quadtree* qt, Point p) {
    if (!contains(qt->boundary, p))
        return 0;

    if (qt->count < CAPACITY) {
        qt->points[qt->count++] = p;
        return 1;
    }

    if (!qt->divided)
        subdivide(qt);

    if (insert(qt->nw, p)) return 1;
    if (insert(qt->ne, p)) return 1;
    if (insert(qt->sw, p)) return 1;
    if (insert(qt->se, p)) return 1;

    return 0;
}

// Check rectangle overlap
int intersects(Boundary a, Boundary b) {
    return !(b.x - b.w > a.x + a.w ||
             b.x + b.w < a.x - a.w ||
             b.y - b.h > a.y + a.h ||
             b.y + b.h < a.y - a.h);
}

// Range query
void query(Quadtree* qt, Boundary range) {
    if (!intersects(qt->boundary, range))
        return;

    for (int i = 0; i < qt->count; i++) {
        if (contains(range, qt->points[i])) {
            printf("(%d, %d)\n", qt->points[i].x, qt->points[i].y);
        }
    }

    if (qt->divided) {
        query(qt->nw, range);
        query(qt->ne, range);
        query(qt->sw, range);
        query(qt->se, range);
    }
}

// Display all points
void display(Quadtree* qt) {
    for (int i = 0; i < qt->count; i++) {
        printf("(%d, %d)\n", qt->points[i].x, qt->points[i].y);
    }

    if (qt->divided) {
        display(qt->nw);
        display(qt->ne);
        display(qt->sw);
        display(qt->se);
    }
}

// Main
int main() {
    Boundary boundary = {0, 0, 50, 50};
    Quadtree* qt = createQuadtree(boundary);

    int choice, x, y;

    while (1) {
        printf("\n--- Quadtree Menu ---\n");
        printf("1. Insert Point\n2. Display Points\n3. Range Query\n4. Exit\n");
        printf("Enter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter x y: ");
                scanf("%d %d", &x, &y);
                insert(qt, (Point){x, y});
                break;

            case 2:
                printf("All Points:\n");
                display(qt);
                break;

            case 3: {
                Boundary range;
                printf("Enter range center (x y): ");
                scanf("%d %d", &range.x, &range.y);
                printf("Enter half width & height: ");
                scanf("%d %d", &range.w, &range.h);

                printf("Points in range:\n");
                query(qt, range);
                break;
            }

            case 4:
                exit(0);

            default:
                printf("Invalid choice!\n");
        }
    }
}

File: Anush/anush sangeeth  s/quadtree.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y


class Rectangle:
    def __init__(self, x, y, w, h):
        self.x = x
        self.y = y
        self.w = w
        self.h = h

    def contains(self, point):
        return (self.x - self.w <= point.x <= self.x + self.w and
                self.y - self.h <= point.y <= self.y + self.h)

    def intersects(self, range_rect):
        return not (range_rect.x - range_rect.w > self.x + self.w or
                    range_rect.x + range_rect.w < self.x - self.w or
                    range_rect.y - range_rect.h > self.y + self.h or
                    range_rect.y + range_rect.h < self.y - self.h)


class QuadTree:
    def __init__(self, boundary, capacity):
        self.boundary = boundary
        self.capacity = capacity
        self.points = []
        self.divided = False

    def subdivide(self):
        x = self.boundary.x
        y = self.boundary.y
        w = self.boundary.w / 2
        h = self.boundary.h / 2

        ne = Rectangle(x + w, y - h, w, h)
        nw = Rectangle(x - w, y - h, w, h)
        se = Rectangle(x + w, y + h, w, h)
        sw = Rectangle(x - w, y + h, w, h)

        self.northeast = QuadTree(ne, self.capacity)
        self.northwest = QuadTree(nw, self.capacity)
        self.southeast = QuadTree(se, self.capacity)
        self.southwest = QuadTree(sw, self.capacity)

        self.divided = True

    def insert(self, point):
        if not self.boundary.contains(point):
            return False

        if len(self.points) < self.capacity:
            self.points.append(point)
            return True

        if not self.divided:
            self.subdivide()

        return (self.northeast.insert(point) or
                self.northwest.insert(point) or
                self.southeast.insert(point) or
                self.southwest.insert(point))

    def query(self, range_rect, found=None):
        if found is None:
            found = []

        if not self.boundary.intersects(range_rect):
            return found

        for p in self.points:
            if range_rect.contains(p):
                found.append(p)

        if self.divided:
            self.northwest.query(range_rect, found)
            self.northeast.query(range_rect, found)
            self.southwest.query(range_rect, found)
            self.southeast.query(range_rect, found)

        return found


if __name__ == "__main__":
    boundary = Rectangle(0, 0, 10, 10)
    qt = QuadTree(boundary, 4)

    points = [Point(1, 2), Point(-3, 4), Point(5, -6), Point(7, 8), Point(0, 0)]

    for p in points:
        qt.insert(p)

    search_area = Rectangle(0, 0, 5, 5)
    result = qt.query(search_area)

    for p in result:
        print(p.x, p.y)
